File: scripts/SSOTokenGenerator.js
var crypto = require('crypto');

// Company info
// go to: https://app.klipfolio.com/settings/single_signon_edit
// to find your companyId and secretKey
// --------------------------------------------------------------
var companyId = ""; // Your Klipfolio Account ID
var sso = "";       // Your SSO Key
var email = "";      // User email in account

var params = {
  "email" : email
}

var salt = sso.concat(companyId);
var hash = crypto.createHash('sha1');
hash.update(salt);
let key = hash.digest().slice(0,16);

var iv = crypto.randomBytes(16);

var cipher = crypto.createCipheriv('aes-128-cbc', key, iv);

var message_buff = new Buffer(JSON.stringify(params));
var data = Buffer.concat([iv, message_buff]);
//Cipher will automatically pad data
var encrypted_buff = cipher.update(data);

var sso_token = Buffer.concat([encrypted_buff, cipher.final()]).toString('base64');

console.log("Test this token: https://klipfolio.github.io/kfapp-custom-sso/index.html\n\n" +
                    "email = " + email + "\n" +
                    "token = " + sso_token + "\n" +
                    "company id = " + companyId);

File: scripts/SSOTokenGenerator.py
#!/usr/bin/python
from Crypto import Random
from Crypto.Cipher import AES
import json
import hashlib
import base64

BLOCK_SIZE = 16
# Company Info
# go to: https://app.klipfolio.com/settings/single_signon_edit
# to find your companyId and secretKey
# --------------------------------------------------------------
email = ''
companyId = ''
sso = ''
salt = sso+companyId

params = {'email':email}
json_string = json.dumps(params, separators=(',',':'))

hash_salt = hashlib.sha1(salt.encode()).digest()[:16]
iv = Random.new().read(AES.block_size)
json_bytes = json_string.encode()
pad = BLOCK_SIZE - len(json_string)%BLOCK_SIZE

aes = AES.new(hash_salt, AES.MODE_CBC, iv)
encrypted_bytes = aes.encrypt( (iv + json_bytes) + pad*chr(pad).encode())
sso_token = base64.b64encode(encrypted_bytes)

print ("Test this token: https://klipfolio.github.io/kfapp-custom-sso/index.html\n\n" + "email = " +email+ "\ntoken = " +sso_token+ "\ncompany id = " +companyId)
